vaadin-combo-box-flow/src/test/java/com/vaadin/flow/component/combobox/dataview/ComboBoxListDataViewTest.java

```diff
@@ -0,0 +1,213 @@
+        ComboBox<String> comboBox = getDefaultLocaleComboBox();
+        setComboBoxTextFilter(comboBox, "ba");
+        items = Arrays.asList("foo", "bar", "banana");
+        ComboBoxListDataView<String> dataView = comboBox.setItems(items);
+        dataView.setFilter(item -> item.length() == 3);
+        Stream<String> allItems = dataView.getItems();
+        Assert.assertArrayEquals("Unexpected data set",
+                new String[] { "bar" }, allItems.toArray());
+    }
+
+    @Test
+    public void getItemCount_noFilters_totalItemsCountObtained() {
+        Assert.assertEquals("Unexpected size for data", items.size(),
+                dataView.getItemCount());
+    }
+
+    @Test
+    public void getItemCount_withFilter_totalItemsCountObtained() {
+        dataView.setFilter(item -> item.equalsIgnoreCase("middle"));
+
+        Assert.assertEquals("Unexpected item count", 1,
+                dataView.getItemCount());
+
+        Assert.assertTrue("Unexpected item", dataView.contains(items.get(1)));
+    }
+
+    @Test
+    public void setIdentifierProvider_customIdentifier_keyMapperUsesIdentifier() {
+        Item first = new Item(1L, "first");
+        Item second = new Item(2L, "middle");
+
+        List<Item> items = new ArrayList<>(Arrays.asList(first, second));
+
+        ComboBox<Item> component = new ComboBox<>();
+
+        DataCommunicator<Item> dataCommunicator = new DataCommunicator<>(
+                (item, jsonObject) -> {
+                }, null, null, component.getElement().getNode());
+
+        ComboBoxListDataView<Item> dataView = new ComboBoxListDataView<>(
+                dataCommunicator, component);
+        DataKeyMapper<Item> keyMapper = dataCommunicator.getKeyMapper();
+        items.forEach(keyMapper::key);
+
+        Assert.assertFalse(keyMapper.has(new Item(1L, "non-present")));
+        dataView.setIdentifierProvider(Item::getId);
+        Assert.assertTrue(keyMapper.has(new Item(1L, "non-present")));
+    }
+
+    @Test
+    public void addItemCountChangeListener_serverSideFilterEnabled() {
+        Assert.assertTrue(Boolean.parseBoolean(getClientSideFilter()));
+
+        dataView.addItemCountChangeListener(event -> {});
+
+        // Trigger refresh all event
+        dataView.setFilter(item -> true);
+
+        Assert.assertFalse(Boolean.parseBoolean(getClientSideFilter()));
+    }
+
+    @Test
+    public void addItemCountChangeListener_removeListeners_serverSideFilterDisabled() {
+        Registration registration1 =
+                dataView.addItemCountChangeListener(event -> {});
+
+        Registration registration2 =
+                dataView.addItemCountChangeListener(event -> {});
+
+        // Trigger ComboBox::refreshAllData()
+        dataView.setFilter(item -> true);
+
+        registration1.remove();
+        Assert.assertFalse(Boolean.parseBoolean(getClientSideFilter()));
+
+        registration2.remove();
+        Assert.assertTrue(Boolean.parseBoolean(getClientSideFilter()));
```

Comment:
I think I commented on this exact same test elsewhere, I don't get how count listener and filtering are related ?

Reply:
This code has been removed.